Add tiflash cdc for performance overview (#11653)
pingcap/docs: performance-tuning-methods.md

- Database Time by SQL Phase: Most time is consumed in the execute phase in green.
- SQL Execute Time Overview: The KV request PessimisticLock shown in red consumes the most time during SQL execution, and the execution time is obviously longer than the total time of KV requests. This is caused by serious lock conflicts in write statements and frequent lock retries prolong `Retried execution time`. Currently, TiDB does not measure `Retried execution time`.

> **Example 5: HTAP CH-Benchmark Load**

Author: ```suggestion
**Example 5: HTAP CH-Benchmark workload**
```

- QPS: Short for Query Per Second. It shows the count of SQL statements executed by the application.
- CPS By Type: Short for Command Per Second. Command indicates MySQL protocol-specific commands. A query statement can be sent to TiDB either by a query command or a prepared statement.
- Queries Using Plan Cache OPS: `avg-hit` is the number of queries using the execution plan cache per second in a TiDB cluster, and `avg-miss` is the number of queries not using the execution plan cache per second in a TiDB cluster.

    `avg-hit + avg-miss` is equal to `StmtExecute`, which is the number of all queries executed per second. When prepared plan cache is enabled in TiDB, the following three scenarios will occur:

>     - No prepared plan cache is hit: `avg-hit` (the number of hits per second) is 0, and `avg-miss` equal to the number of `StmtExecute` commands per second. The possible reasons include:

Author: ```suggestion
    - No prepared plan cache is hit: `avg-hit` (the number of hits per second) is 0, and `avg-miss` is equal to the number of `StmtExecute` commands per second. The possible reasons include:
```

        - The application is using the query interface.
        - The cached plans are cleaned up because the application calls the `StmtClose` command after each `StmtExecute` execution.

>         - All statements executed by `StmtExecute` do not meet the [cache conditions](sql-prepared-plan-cache.md) so the execution plan cache cannot be hit.

Author: ```suggestion
        - All statements executed by `StmtExecute` do not meet the [cache conditions](/sql-prepared-plan-cache.md) so the execution plan cache cannot be hit.
```

    - All prepared plan cache is hit: `avg-hit` (the number of hits per second) is equal to the number of `StmtExecute` commands per second, and `avg-miss` (the number without hits per second) is 0.

>     - Some prepared plan cache is hit: `avg-hit` (the number of hits per second) is fewer than the number of `StmtExecute` commands per second. Prepared plan cache has known limitations: for example, it does not support subqueries, so SQL statements with subqueries cannot use prepared plan cache.

Reviewer: ```suggestion
    - Some prepared plan cache is hit: `avg-hit` (the number of hits per second) is fewer than the number of `StmtExecute` commands per second. Prepared plan cache has known limitations. For example, it does not support subqueries, so SQL statements with subqueries cannot use prepared plan cache.
```

- In the CPS By Type panel, the number of `StmtPrepare` commands per second is much greater than that of `StmtClose` per second, which indicates that an object leak occurs in the application for prepared statements.
- In the Queries Using Plan Cache OPS panel, `avg-miss` is almost equal to `StmtExecute` in the CPS By Type panel, which indicates that almost all SQL executions miss the execution plan cache.

#### KV/TSO Request OPS and KV Request Time By Source

- In the KV/TSO Request OPS panel, you can view the statistics of KV and TSO requests per second. Among the statistics, `kv request total` represents the sum of all requests from TiDB to TiKV. By observing the types of requests from TiDB to PD and TiKV, you can get an idea of the workload profile within the cluster.
- In the KV Request Time By Source panel, you can view the time ratio of each KV request type and all request sources.
    - kv request total time: The total time of processing KV and TiFlash requests per second.

>     - Each KV request and the corresponding request source form a columnar stacked graph, in which `external` identifies normal business requests and `internal` identifies internal activity requests (such as DDL and auto analyze requests).

Reviewer: ```suggestion
    - Each KV request and the corresponding request source form a stacked bar chart, in which `external` identifies normal business requests and `internal` identifies internal activity requests (such as DDL and auto analyze requests).
```

> 
> In the KV Request Time By Source panel, you can view the total number of connections and the number of connections per TiDB. 

Author: ```suggestion
```

**Example 1: Busy workload**

![TPC-C](/media/performance/tpcc_source_sql.png)

In this TPC-C workload:

> - The total number of KV requests per second is 7,970. The top request types are `Prewrite`, `Commit`, `PessimisticsLock`, and `BatchGet` in order of number of requests.

Reviewer: The Chinese version is "79.7K".
```suggestion
- The total number of KV requests per second is 79,700. The top request types are `Prewrite`, `Commit`, `PessimisticsLock`, and `BatchGet` in order of number of requests.
```

![OLTP-Prepared](/media/performance/oltp_prepared_statement_no_plan_cache.png)

**Example 4: Prepared statements have a resource leak**

> The number of `StmtExecute` commands per second is much greater than that of `StmtClose` per second, which indicates that the application has a object leak for prepared statements.

Author: ```suggestion
The number of `StmtPrepare` commands per second is much greater than that of `StmtClose` per second, which indicates that the application has an object leak for prepared statements.
```

![OLTP-Query](/media/performance/prepared_statement_leaking.png)

- In the QPS panel, the red bold line indicates the number of failed queries, and the Y axis on the right indicates the coordinate value of the number. In this example, the number of failed quries per second is 74.6.

> - In the CPS By Type panel, the number of `StmtExecute` commands per second is much greater than that of `StmtClose` per second, which indicates that a object leak occurs in the application for prepared statements.

Author: ```suggestion
- In the CPS By Type panel, the number of `StmtPrepare` commands per second is much greater than that of `StmtClose` per second, which indicates that an object leak occurs in the application for prepared statements.
```

> - In the Queries Using Plan Cache OPS panel, `avg-miss` is almost equal to `StmtExecute` in the CPS By Type panel, which indicates that almost all SQL executions miss the execution plan cache.

Author: ```suggestion
- In the Queries Using Plan Cache OPS panel, `avg-miss` is almost equal to `StmtExecute`, which indicates that almost all SQL executions miss the execution plan cache.
```

#### KV/TSO Request OPS and KV Request Time By Source

- In the KV/TSO Request OPS panel, you can view the statistics of KV and TSO requests per second. Among the statistics, `kv request total` represents the sum of all requests from TiDB to TiKV. By observing the types of requests from TiDB to PD and TiKV, you can get an idea of the workload profile within the cluster.

> - In the KV Request Time By Source panel, you can view the time ratio of each KV request and all request sources.

Author: ```suggestion
- In the KV Request Time By Source panel, you can view the time ratio of each KV request type and all request sources.
```

    - kv request total time: The total time of processing KV and TiFlash requests per second.
    - Each KV request and the corresponding request source form a columnar stacked graph, in which `external` identifies normal business requests and `internal` identifies internal activity requests (such as DDL and auto analyze requests).

In the KV Request Time By Source panel, you can view the total number of connections and the number of connections per TiDB. 

> 
> The counts help you determine whether the total number of connections is normal and the number of connections per TiDB is even. `active connections` records the number of active connections, which is equal to the database time per second.

Author: ```suggestion
```

**Example 1: Busy workload**

![TPC-C](/media/performance/tpcc_source_sql.png)

In this TPC-C workload:

- The total number of KV requests per second is 7,970. The top request types are `Prewrite`, `Commit`, `PessimisticsLock`, and `BatchGet` in order of number of requests.
- Most of the KV processing time is spent on `Commit-external_Commit` and `Prewrite-external_Commit`, which indicates that the most time-consuming KV requests are `Commit` and `Prewrite` from external commit statements.

**Example 2: Analyze workload**

![OLTP](/media/performance/internal_stats.png)

> In this workload:

Author: ```suggestion
In this workload, only `ANALYZE` statements are running in the cluster:
```

> - The total number of KV requests per second is 3050 and the number of Cop requests per second is 1100.

Author: ```suggestion
- The total number of KV requests per second is 35.5 and the number of Cop requests per second is 9.3.
```

> - Most of the KV processing time is spent on `Cop-internal_stats`, which indicates that the most time-consuming KV request is `Cop` from internal analyze operations.

Author: ```suggestion
- Most of the KV processing time is spent on `Cop-internal_stats`, which indicates that the most time-consuming KV request is `Cop` from internal `ANALYZE` operations.
```

- The network latency from the application server to the database is high. For example, the network latency is high because in public-cloud deployments the applications and the TiDB cluster are not in the same region, or the dns workload balancer and the TiDB cluster are not in the same region.
- The bottleneck is in client applications. The application server's CPU cores and Numa resources cannot be fully utilized. For example, only one JVM is used to establish thousands of JDBC connections to TiDB.

> In the Connection Count panel, you can check the total number of connections and also the number of connections on each TiDB node, which helps you determine whether the total number of connections is normal and whether the number of connections on each TiDB node is unbalanced. `active connections` indicates the number of active connections, which is equal to the database time per second. The Y axis on the right indicates the number of `disconnection/s` per second in a cluster, which can be used to determine whether the application uses short connections.

Reviewer: ```suggestion
In the Connection Count panel, you can check the total number of connections and also the number of connections on each TiDB node, which helps you determine whether the total number of connections is normal and whether the number of connections on each TiDB node is unbalanced. `active connections` indicates the number of active connections, which is equal to the database time per second. The Y axis on the right (`disconnection/s`) indicates the number of disconnections per second in a cluster, which can be used to determine whether the application uses short connections.
```

> In the Connection Count panel, you can check the total number of connections and also the number of connections on each TiDB node, which helps you determine whether the total number of connections is normal and whether the number of connections on each TiDB node is unbalanced. `active connections` indicates the number of active connections, which is equal to the database time per second. The Y axis on the right is for `disconnection/s`, which indicates the number of disconnections per second in a cluster, which can be used to determine whether the application uses short connections.

Author: ```suggestion
In the Connection Count panel, you can check the total number of connections and also the number of connections on each TiDB node, which helps you determine whether the total number of connections is normal and whether the number of connections on each TiDB node is unbalanced. `active connections` indicates the number of active connections, which is equal to the database time per second. The Y axis on the right indicates the number of `disconnection/s` per second in a cluster, which can be used to determine whether the application uses short connections.
```

**Example 1: The number of disconnection/s is too high**

![high disconnection/s](/media/performance/high_disconnections.png)

In this workload:

- The average latency and P99 latency of all SQL statements are 10.8 ms and 84.1 ms, respectively.
- The average connection idle time in transactions `avg-in-txn` is 9.4 ms.

> - The total number of connections to the cluster is 3700, and the number of connections to each TiDB node is 1800. The average number of active connections is 40.3, which indicates that most of connections are idle. The average number of `disonnnection/s` is 55.8, which indicates that the application is connecting and disconnecting frequently. The behavior of short connections will have a certain impact on TiDB resources and response time.

Reviewer: ```suggestion
- The total number of connections to the cluster is 3,700, and the number of connections to each TiDB node is 1,800. The average number of active connections is 40.3, which indicates that most of the connections are idle. The average number of `disonnnection/s` is 55.8, which indicates that the application is connecting and disconnecting frequently. The behavior of short connections will have a certain impact on TiDB resources and response time.
```

- The network latency from the application server to the database is high. For example, the network latency is high because in public-cloud deployments the applications and the TiDB cluster are not in the same region, or the dns workload balancer and the TiDB cluster are not in the same region.
- The bottleneck is in client applications. The application server's CPU cores and Numa resources cannot be fully utilized. For example, only one JVM is used to establish thousands of JDBC connections to TiDB.

> In the Connection Count panel, you can check the total number of connections and also the number of connections on each TiDB node, which helps you evaluate whether the total number of connections is normal and whether the number of connections on each TiDB node is unbalanced. `active connections` indicates the number of active connections, which is equal to the database time per second. The Y axis on the right indicates `disconnection/s`, which indicates the number of disconnections per second in a cluster, which can be used to determine whether the application uses short connections.

Author: ```suggestion
In the Connection Count panel, you can check the total number of connections and also the number of connections on each TiDB node, which helps you determine whether the total number of connections is normal and whether the number of connections on each TiDB node is unbalanced. `active connections` indicates the number of active connections, which is equal to the database time per second. The Y axis on the right is for `disconnection/s`, which indicates the number of disconnections per second in a cluster, which can be used to determine whether the application uses short connections.
```

The average query latency is significantly greater than `avg-in-txn`, which means the main bottleneck in transactions is inside the database.

> **Example 3: The bottleneck of user response time is not in TiDB**

Reviewer: To keep consistent with example 2
```suggestion
**Example 3: TiDB is not the bottleneck of user response time**
```
